File: c/implementations/medium/sharpen.c
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>

void apply_sharpen(uint8_t *pixels, uint8_t *output, size_t width,
                   size_t height) {
  size_t total_bytes = width * height * PX_SIZE;
  for (size_t i = 0; i < total_bytes; i++) {
    output[i] = pixels[i];
  }

  int row_bytes = width * PX_SIZE;
  for (size_t y = 1; y < height - 1; y++) {
    size_t idx = (y * width + 1) * PX_SIZE;

    for (size_t x = 1; x < width - 1; x++, idx += PX_SIZE) {

      // c = 0 (Red), 1 (Green), 2 (Blue)
      for (int c = 0; c < 3; c++) {
        // kernel
        int sum = pixels[idx - row_bytes + c] * -1 + // top
                  pixels[idx - PX_SIZE + c] * -1 +   // left
                  pixels[idx + c] * 5 +              // center
                  pixels[idx + PX_SIZE + c] * -1 +   // right
                  pixels[idx + row_bytes + c] * -1;  // bottom

        output[idx + c] = clamp_to_u8_int(sum);
      }
    }
  }
}

File: src/shared/lib/image-processing/medium/sharpen.ts
import type { FilterProcessFn } from "../types";

export const applySharpen: FilterProcessFn = (pixels, width, height) => {
  const output = new Uint8ClampedArray(pixels.length);
  const rowBytes = width * 4;

  for (let y = 1; y < height - 1; y++) {
    let idx = (y * width + 1) * 4;
    for (let x = 1; x < width - 1; x++, idx += 4) {
      // c = 0 (Red), 1 (Green), 2 (Blue)
      for (let c = 0; c < 3; c++) {
        // kernel: [0, -1, 0, -1, 5, -1, 0, -1, 0]
        const sum =
          pixels[idx - rowBytes + c] * -1 + // top
          pixels[idx - 4 + c] * -1 + // left
          pixels[idx + c] * 5 + // center
          pixels[idx + 4 + c] * -1 + // right
          pixels[idx + rowBytes + c] * -1; // bottom

        // inline clamp
        output[idx + c] = sum < 0 ? 0 : sum > 255 ? 255 : sum;
      }
      output[idx + 3] = pixels[idx + 3];
    }
  }
  return output;
};
